### PR TITLE
[TDV-717] New managers

### PR DESCRIPTION
**Pull request by ardier16**
+ Wallets manager
+ Factors manager
+ Signers manager

### CODE REVIEW

File: src/api2/middlewares/change-signers.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import { TransactionBuilder } from '../../base/transaction_builder'
```
Comment (quinn71) [inline src/api2/middlewares/change-signers.js]: it's not middleware, move it somewhere near manager

File: src/api2/wallets-manager.js [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+            attributes: {
+              account_id: encryptedSecondFactorWallet.accountId,
+              keychain_data: encryptedSecondFactorWallet.keychainData,
+              salt: encryptedSecondFactorWallet.salt
+            }
+          }
+        ]
+      })
+    } finally {
+      this._apiCaller.useWallet(oldWallet)
```
Comment (quinn71) [inline src/api2/wallets-manager.js]: mb better to make `putWithSignature` also accept the wallet as a 3rd param?

Comment (quinn71) [reply src/api2/wallets-manager.js]: with the default value

Comment (quinn71) [reply src/api2/wallets-manager.js]: @max-nes

File: src/api2/wallets-manager.js [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+    return wallet.accountId
+  }
+
+  async _getDefaultSignerRole () {
+    const endpoint = `/v3/key_values/${DEFAULT_SIGNER_ROLE_KEY}`
+    const { data } = await this._apiCaller.get(endpoint)
+
+    return String(data.value.u32)
+  }
+}
```
Comment (quinn71) [inline src/api2/wallets-manager.js]: it's looks very big, I suggest extracting `FactorsManager` and `SignerManager` from here

Comment (ardier16) [reply src/api2/wallets-manager.js]: Extracted

File: src/api2/middlewares/change-signers.js (hunk 1 above)
Comment (ardier16) [reply src/api2/middlewares/change-signers.js]: Moved to `SignersManager` class

File: src/api2/wallets-manager.js (hunk 2 above)
Comment (ardier16) [reply src/api2/wallets-manager.js]: Discussed IRL

File: src/api2/index.js [hunk 4]
```diff
@@ -1,2 +1,3 @@
 export { ApiCaller } from './api-caller'
+export { WalletsManager, FactorsManager } from './managers'
```
Comment (max-nes) [inline src/api2/index.js]: missing signers manager

Comment (ardier16) [reply src/api2/index.js]: Added

File: src/api2/managers/signers-manager.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
+import { TransactionBuilder } from '../../base/transaction_builder'
+import { ManageSignerBuilder } from '../../base/operations/manage_signer_builder'
+
+import { NotFoundError } from '../../errors'
+
+const RECOVERY_SIGNER_IDENTITY = 1
```
Comment (quinn71) [inline src/api2/managers/signers-manager.js]: `RECOVERY_SIGNER_ROLE_ID`

File: src/api2/managers/signers-manager.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      operations.push(...removeSignerOps)
+    }
+
+    return operations
+  }
+
+  async _getNonRecoverySigners (accountId) {
+    const signers = await this._getSigners(accountId)
+    return signers
+      .filter(signer => signer.identity !== RECOVERY_SIGNER_IDENTITY)
```
Comment (quinn71) [inline src/api2/managers/signers-manager.js]: role ID, not identity

File: src/api2/managers/signers-manager.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+      return signers
+    } catch (err) {
+      if (err instanceof NotFoundError) {
+        return []
+      }
+    }
+  }
+
+  async _getCreateSignerOp (newAccountId) {
```
Comment (quinn71) [inline src/api2/managers/signers-manager.js]: a bit confusing that above you use `get` for the method that performs http request, here you use it for method it doesn't. use smth like `make` for methods that will return you operations

File: src/api2/managers/signers-manager.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  async _getSigners (accountId) {
+    try {
+      const endpoint = `/v3/accounts/${accountId}/signers`
+      const { data: signers } = await this._apiCaller.get(endpoint)
+
+      return signers
+    } catch (err) {
+      if (err instanceof NotFoundError) {
```
Comment (quinn71) [inline src/api2/managers/signers-manager.js]: collection endpoints shouldn't return you 404, if the collection is empty

Comment (quinn71) [reply src/api2/managers/signers-manager.js]: if it does, pls tell backend team about it

File: src/api2/managers/wallets-manager.spec.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
+import sinon from 'sinon'
```
Comment (quinn71) [inline src/api2/managers/wallets-manager.spec.js]: all tests here check more than described

File: src/api2/managers/factors-manager.spec.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    })
+
+    describe('verifyTotpFactor', () => {
+      it('gets attributes from error instance and sends PUT verification request', async () => {
+        sandbox.stub(factorsManagerInstance._apiCaller, 'put')
+          .withArgs('/wallets/some-wallet-id/factors/1/verification')
+          .resolves()
+
+        await factorsManagerInstance.verifyTotpFactor(
+          {
```
[the remark this answers is not in the document]
Comment (ardier16) [reply src/api2/managers/factors-manager.spec.js]: Passed

File: src/api2/managers/factors-manager.spec.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+          .to.have.been.calledOnceWith(
+            '/wallets/some-wallet-id/factors/1/verification'
+          )
+      })
+    })
+
+    describe('verifyTotpFactor', () => {
+      it('gets attributes from error instance and sends PUT verification request', async () => {
+        sandbox.stub(factorsManagerInstance._apiCaller, 'put')
+          .withArgs('/wallets/some-wallet-id/factors/1/verification')
```
[the remark this answers is not in the document]
Comment (ardier16) [reply src/api2/managers/factors-manager.spec.js]: Updated

File: src/api2/managers/factors-manager.spec.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        })
+        expect(factorsManagerInstance._apiCaller.put)
+          .to.have.been.calledOnceWith(
+            '/wallets/some-wallet-id/factors/1/verification'
+          )
+      })
+    })
+
+    describe('verifyTotpFactor', () => {
+      it('gets attributes from error instance and sends PUT verification request', async () => {
```
[the remark this answers is not in the document]
Comment (ardier16) [reply src/api2/managers/factors-manager.spec.js]: Updated

File: src/api2/managers/factors-manager.spec.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        'GBUQDWXPPEFREJPI45CUPACMY6AQINP4DQ2DFXAF6YISPF3C4FFJ3U5S',
+        'some-wallet-id'
+      )
+
+      beforeEach(() => {
+        factorsManagerInstance._apiCaller.useWallet(mockedWallet)
+      })
+
+      it('gets attributes from error instance and sends PUT verification request', async () => {
+        const keychainData = 'eyJJViI6IkFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFEQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQkFBQUFBQUFBQUFFQUFBQUFBQUFBQUFBQUFBREFBQUFCQUFBQUFRQUFBQUlBQUFBQndBQUFBUUFBQUFBIiwiY2lwaGVyVGV4dCI6Ii90enkreXV6ajdtdEEyTGlVQjFsQ013NWlPVUtjL2owOHVHcW9iK1ZsWk9XaEdHVWlLSUpmSWU0clY1c3JZMzVNRi83M1N4NTd3aUR1MFA0K2h0NjRHYnIzNFV6MHpZMUlZVGVUd2REbWVHZ1NBeVNQTGloS0c3OHNxY1NsV2owQmpxMU13ZVV3SEpXM1ZIMDFSeGlTUk5oSmlFcEtsUDdzOFhwZUhDNDlyM1RWOW9RcS9CcWJMeStERkZHRHNGWFFMenhBYnp2c0VxWGlnPT0iLCJjaXBoZXJOYW1lIjoiYWVzIiwibW9kZU5hbWUiOiJnY20ifQ=='
```
[the remark this answers is not in the document]
Comment (ardier16) [reply src/api2/managers/factors-manager.spec.js]: Updated, use `some-keychain-data` for better readability

File: src/api2/managers/signers-manager.js (hunk 5 above)
Comment (ardier16) [reply src/api2/managers/signers-manager.js]: Renamed

File: src/api2/managers/signers-manager.js (hunk 6 above)
Comment (ardier16) [reply src/api2/managers/signers-manager.js]: Fixed

File: src/api2/managers/signers-manager.js (hunk 7 above)
Comment (ardier16) [reply src/api2/managers/signers-manager.js]: Renamed

File: src/api2/managers/signers-manager.js (hunk 8 above)
Comment (ardier16) [reply src/api2/managers/signers-manager.js]: Yes, it returns empty array. Fixed

File: src/api2/managers/wallets-manager.spec.js (hunk 9 above)
Comment (ardier16) [reply src/api2/managers/wallets-manager.spec.js]: Updated

**Pull request merged**